File: CommunicateIntentLH/Greeter.cs
namespace CommunicateIntentLH;

internal class Greeter
{
    private readonly IDateTimeProvider _datetimeProvider;

    public Greeter(IDateTimeProvider datetimeProvider)
    {
        _datetimeProvider = datetimeProvider;
    }
    
    public string Greet(string name)
    {
        var hour = _datetimeProvider.GetDateTime().Hour;
        string template = string.Empty;
        if (hour.IsBetween(7, 13))
        {
            template = "Good morning {0}";
        }
        if (hour.IsBetween(14,19))
        {
            template = "Good afternoon {0}";
        }
        if (hour.IsBetween(20, 6))
        {
            template = "Good night {0}";
        }
        return String.Format(template, name);
    }
}

File: kotlin/src/main/kotlin/Greeter.kt
internal class Greeter(private val dateTimeProvider: IDateTimeProvider) {
    
    fun greet(name: String?): String {
        requireNotNull(name) { "name cannot be null" }
        
        val hour = dateTimeProvider.getDateTime().hour
        val template = when {
            hour in 7..13 -> "Good morning %s"
            hour in 14..19 -> "Good afternoon %s"
            hour in 20..23 || hour in 0..6 -> "Good night %s"
            else -> ""
        }
        return template.format(name)
    }
}